File: core/src/ru/myitschool/dcrawler/event/EventController.java
package ru.myitschool.dcrawler.event;

import com.badlogic.gdx.utils.Array;

import java.util.HashMap;

public class EventController {

    private static Array<EventAction> beforeActions = new Array<>();
    private static Array<EventAction> afterActions = new Array<>();

    private static HashMap<String, EventSlot> events = new HashMap<>();

    public static HashMap<String, Object> callEvent(String eventCode, HashMap<String, Object> args){
        beforeEvent(eventCode);
        HashMap<String, Object> result = event(eventCode, args);
        afterEvent(eventCode);
        return result;
    }

    private static void beforeEvent(String eventCode){
        for (EventAction action : beforeActions){
            action.act(eventCode);
        }
    }

    private static HashMap<String, Object> event(String eventCode, HashMap<String, Object> args){
        return events.get(eventCode).fire(eventCode, args);
    }

    private static void afterEvent(String eventCode){
        for (EventAction action : afterActions){
            action.act(eventCode);
        }
    }

    public static EventSlot registerEvent(String eventCode) throws Exception {
        if (!events.containsKey(eventCode)) {
            EventSlot slot = new EventSlot();
            events.put(eventCode, slot);
            return slot;
        } else {
            throw new Exception("Event with code " + eventCode + " already registered");
        }
    }

    public static void addActionBeforeEvent(EventAction action){
        beforeActions.add(action);
    }

    public static void addActionAfterEvent(EventAction action){
        afterActions.add(action);
    }
}

File: core/src/com/wolg_vlad/dcrawler/event/EventController.kt
package com.wolg_vlad.dcrawler.event

import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.utils.throwable.EventAlreadyRegisteredError
import com.wolg_vlad.dcrawler.utils.throwable.EventNotFoundError
import java.util.*

object EventController {
    private val beforeActions = Array<EventAction>()
    private val afterActions = Array<EventAction>()
    private val events = HashMap<String, EventSlot?>()
    fun callEvent(eventCode: String, args: HashMap<String, Any>): HashMap<String, Any> {
        beforeEvent(eventCode)
        val result = event(eventCode, args)
        afterEvent(eventCode)
        return result
    }

    private fun beforeEvent(eventCode: String) {
        for (action in beforeActions) {
            action.act(eventCode)
        }
    }

    private fun event(eventCode: String, args: HashMap<String, Any>): HashMap<String, Any> {
        if (events[eventCode] == null) {
            throw EventNotFoundError(eventCode)
        }
        return events[eventCode]!!.fire(eventCode, args)
    }

    private fun afterEvent(eventCode: String) {
        for (action in afterActions) {
            action.act(eventCode)
        }
    }

    @Throws(Exception::class)
    fun registerEvent(eventCode: String, listener: EventListener): EventSlot {
        return if (!events.containsKey(eventCode)) {
            val slot = EventSlot(listener)
            events[eventCode] = slot
            slot
        } else {
            throw EventAlreadyRegisteredError(eventCode)
        }
    }

    fun addActionBeforeEvent(action: EventAction) {
        beforeActions.add(action)
    }

    fun addActionAfterEvent(action: EventAction) {
        afterActions.add(action)
    }
}